### ISSUE TITLE
`promtool test rules` fails when /tmp is not writable

### ISSUE THREAD

**Issue by orfeas-k** [prometheus/prometheus#13703]
### What did you do?

Tried to run `promtool` from the prometheus docker and it fails while it works for the same files

```
$ promtool test rules tests/unit/promtool/test_unit_available.yaml                                                                                                                         Unit Testing:  tests/unit/promtool/test_unit_available.yaml
  SUCCESS

$ docker run -v $PWD:/tmp --entrypoint "/bin/promtool" prom/prometheus:latest  test rules /tmp/tests/unit/promtool/test_unit_available.yaml                                              
Unit Testing:  /tmp/tests/unit/promtool/test_unit_available.yaml
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x18 pc=0x19f3f48]

goroutine 1 [running]:
github.com/prometheus/prometheus/promql.(*LazyLoader).Errorf(0xc00067a000?, {0x34cdded?, 0xc000111c20?}, {0xc0006d7e10?, 0x2fcd000?, 0x56bb9a0?})
	<autogenerated>:1 +0x28
github.com/stretchr/testify/assert.Fail({0x7fcef76a0ba0, 0xc000111c20}, {0xc0000544b0, 0x4f}, {0xc000a40060, 0x1, 0x1})
	/go/pkg/mod/github.com/stretchr/testify@v1.8.4/assert/assertions.go:333 +0x350
github.com/stretchr/testify/assert.NoError({0x7fcef76a0ba0, 0xc000111c20}, {0x3bb7500, 0xc0007d1830}, {0xc000a40060, 0x1, 0x1})
	/go/pkg/mod/github.com/stretchr/testify@v1.8.4/assert/assertions.go:1495 +0xfc
github.com/stretchr/testify/require.NoError({0x7fcef76a0b58, 0xc000111c20}, {0x3bb7500, 0xc0007d1830}, {0xc000a40060, 0x1, 0x1})
	/go/pkg/mod/github.com/stretchr/testify@v1.8.4/require/require.go:1357 +0x8f
github.com/prometheus/prometheus/util/teststorage.New({0x3bc6af8?, 0xc000111c20})
	/app/util/teststorage/storage.go:34 +0xc5
github.com/prometheus/prometheus/promql.(*LazyLoader).clear(0xc000111c20)
	/app/promql/test.go:772 +0xd6
github.com/prometheus/prometheus/promql.NewLazyLoader({0x0?, 0x0}, {0xc000492070, 0x68}, {0x60?, 0xac?})
	/app/promql/test.go:736 +0xa9
main.(*testGroup).test(0xc000a40fa0, 0xdf8475800, 0xc0005930e0?, {0xe0?, 0x8d?}, 0x0, {0xc0006d7d10, 0x1, 0x1})
	/app/cmd/promtool/unittest.go:180 +0xa5
main.ruleUnitTest({0x7ffd516b4f55, 0x31}, {0xa?, 0x0?}, 0xc000a41188?, 0x4c?)
	/app/cmd/promtool/unittest.go:114 +0x58c
main.RulesUnitTest({0x0?, 0x0?}, {0x0?, 0x3ba9140?, 0x1?}, 0x0?, {0xc0006d76e0?, 0x1, 0x0?})
	/app/cmd/promtool/unittest.go:54 +0x131
main.main()
	/app/cmd/promtool/main.go:373 +0xb278
```

### What did you expect to see?

Same output as when I download promtool binary e.g.
```
Unit Testing:  tests/unit/promtool/test_unit_available.yaml
  SUCCESS
```

### What did you see instead? Under which circumstances?

```
Unit Testing:  /tmp/tests/unit/promtool/test_unit_available.yaml
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x18 pc=0x19f3f48]

goroutine 1 [running]:
github.com/prometheus/prometheus/promql.(*LazyLoader).Errorf(0xc00067a000?, {0x34cdded?, 0xc000111c20?}, {0xc0006d7e10?, 0x2fcd000?, 0x56bb9a0?})
	<autogenerated>:1 +0x28
github.com/stretchr/testify/assert.Fail({0x7fcef76a0ba0, 0xc000111c20}, {0xc0000544b0, 0x4f}, {0xc000a40060, 0x1, 0x1})
	/go/pkg/mod/github.com/stretchr/testify@v1.8.4/assert/assertions.go:333 +0x350
github.com/stretchr/testify/assert.NoError({0x7fcef76a0ba0, 0xc000111c20}, {0x3bb7500, 0xc0007d1830}, {0xc000a40060, 0x1, 0x1})
	/go/pkg/mod/github.com/stretchr/testify@v1.8.4/assert/assertions.go:1495 +0xfc
github.com/stretchr/testify/require.NoError({0x7fcef76a0b58, 0xc000111c20}, {0x3bb7500, 0xc0007d1830}, {0xc000a40060, 0x1, 0x1})
	/go/pkg/mod/github.com/stretchr/testify@v1.8.4/require/require.go:1357 +0x8f
github.com/prometheus/prometheus/util/teststorage.New({0x3bc6af8?, 0xc000111c20})
	/app/util/teststorage/storage.go:34 +0xc5
github.com/prometheus/prometheus/promql.(*LazyLoader).clear(0xc000111c20)
	/app/promql/test.go:772 +0xd6
github.com/prometheus/prometheus/promql.NewLazyLoader({0x0?, 0x0}, {0xc000492070, 0x68}, {0x60?, 0xac?})
	/app/promql/test.go:736 +0xa9
main.(*testGroup).test(0xc000a40fa0, 0xdf8475800, 0xc0005930e0?, {0xe0?, 0x8d?}, 0x0, {0xc0006d7d10, 0x1, 0x1})
	/app/cmd/promtool/unittest.go:180 +0xa5
main.ruleUnitTest({0x7ffd516b4f55, 0x31}, {0xa?, 0x0?}, 0xc000a41188?, 0x4c?)
	/app/cmd/promtool/unittest.go:114 +0x58c
main.RulesUnitTest({0x0?, 0x0?}, {0x0?, 0x3ba9140?, 0x1?}, 0x0?, {0xc0006d76e0?, 0x1, 0x0?})
	/app/cmd/promtool/unittest.go:54 +0x131
main.main()
	/app/cmd/promtool/main.go:373 +0xb278
```

Other environment information:

```
$ docker version                                                                                                                            
Client:
 Version:           24.0.5
 API version:       1.43
 Go version:        go1.20.8
 Git commit:        ced0996
 Built:             Sat Oct  7 00:14:00 2023
 OS/Arch:           linux/amd64
 Context:           default

Server:
 Engine:
  Version:          24.0.5
  API version:      1.43 (minimum version 1.12)
  Go version:       go1.20.8
  Git commit:       a61e2b4
  Built:            Sat Oct  7 00:14:30 2023
  OS/Arch:          linux/amd64
  Experimental:     false
 containerd:
  Version:          v1.6.21
  GitCommit:        3dce8eb055cbb6872793272b4f20ed16117344f8
 runc:
  Version:          1.1.7
  GitCommit:        
 docker-init:
  Version:          0.19.0
  GitCommit:        de40ad0
```

### System information

Linux 6.5.0-21-generic x86_64

### Prometheus version

```text
2.50.0 but this happens with 2.20.0 image,2.30.0, 2.40.0
```

**Comment by beorn7** [comment]
Weird. @dgl any idea why this is apparently only happening with `promtool` from the docker image?

**Comment by dgl** [comment]
Fun bug... the problem is `docker run` with `-v $PWD:/tmp` -- this is making /tmp be mounted as whatever the working directory is, which likely isn't set up to be world writable like /tmp is expected to be. This means `os.MkdirTemp` fails (as the Docker image runs as `nobody`). I clearly need to fix the error reporting though, it shouldn't be using testify on this code path.

tl;dr: Rather than using `/tmp`, use something without a special meaning like `/data`.

**Comment by orfeas-k** [comment]
Good catch actually, it never occured to me that mounting `tmp` could create an issue. I 'll make sure to avoid this practice in the future. 

Thanks for the workaround. I 'll leave this open since it's something you need to fix.
